renku/core/models/provenance/activity.py

```diff
@@ -0,0 +1,417 @@
+        host = get_host(client)
+        return urljoin(f"https://{host}", pathlib.posixpath.join("activities", str(uuid.uuid4())))
+
+    @classmethod
+    def from_jsonld(cls, data):
+        """Create an instance from JSON-LD."""
+        if isinstance(data, cls):
+            return data
+        if not isinstance(data, list):
+            raise ValueError(data)
+
+        return ActivitySchema(flattened=True).load(data)
+
+    def to_jsonld(self):
+        """Create JSON-LD."""
+        return ActivitySchema(flattened=True).dump(self)
+
+
+def _convert_qualified_usage(qualified_usage: List[Usage], activity_id, client) -> List[Usage]:
+    """Convert a qualified Usages."""
+    usages = []
+
+    for usage in qualified_usage:
+        commit_sha = _extract_commit_sha(entity_id=usage.entity._id)
+        entity = _convert_usage_entity(usage.entity, commit_sha, activity_id, client)
+        assert entity, f"Top entity was not found for Usage: {usage._id}, {usage.entity.path}"
+
+        id_ = pathlib.posixpath.join("usage", str(uuid.uuid4()), entity.checksum, quote(entity.path))
+        usage_id = f"{activity_id}/{id_}"
+        duplicates_found = any([u for u in usages if u._id == usage_id])
+        assert not duplicates_found
+
+        new_usage = Usage(id=usage_id, entity=entity, role=usage.role)
+        usages.append(new_usage)
+
+    return usages
+
+
+def _convert_generated(generated: List[Generation], activity_id, client) -> List[Generation]:
+    """Convert Generations."""
+    generations = []
+
+    for generation in generated:
+        commit_sha = _extract_commit_sha(entity_id=generation.entity._id)
+        entity = _convert_generation_entity(generation.entity, commit_sha, activity_id, client)
+        assert entity, f"Root entity was not found for Generation: {generation._id}"
+
+        quoted_path = quote(entity.path)
+        id_ = pathlib.posixpath.join("generation", str(uuid.uuid4()), entity.checksum, quoted_path)
+        generation_id = f"{activity_id}/{id_}"
+
+        duplicates_found = any([g for g in generations if g._id == generation_id])
+        assert not duplicates_found
+
+        new_generation = Generation(id=generation_id, entity=entity, role=generation.role)
+        generations.append(new_generation)
+
+    return generations
+
+
+def _convert_usage_entity(entity: Entity, revision, activity_id, client) -> Union[Entity, None]:
+    """Convert an Entity to one with proper metadata.
+
+    For Collections, add members that are modified in the same commit or before the revision.
+    """
+    assert isinstance(entity, Entity)
+
+    checksum = _get_object_hash(revision=revision, path=entity.path, client=client)
+    if not checksum:
+        return None
+
+    id_ = _generate_entity_id(entity_checksum=checksum, path=entity.path, activity_id=activity_id)
+
+    if isinstance(entity, Collection):
+        new_entity = Collection(id=id_, checksum=checksum, path=entity.path)  # , project=entity._project)
```

Comment:
ATM, entities don't have their `project` property set because due to dynamic id assignment we will end up with have multiple project ids for the same project. I commented these out, but we should add them back when we have fixed non-dynamic project ids. Having this in mind, it's ok to remove commented out project assignments.